arcgis-ios-sdk-samples/Route & Directions/Navigate route/NavigateRouteViewController.swift

```diff
@@ -0,0 +1,296 @@
+                    mapView.setViewpoint(viewPoint, completion: nil)
+                } else {
+                    mapView.setViewpointGeometry(routeGeometry) { [weak self] _ in
+                        // Get the initial zoomed view point.
+                        self?.defaultViewPoint = self?.mapView.currentViewpoint(with: .centerAndScale)
+                    }
+                }
+            }
+        }
+    }
+    
+    /// Resets to the starting location for location display.
+    func resetToStartingLocation() {
+        guard let initialLocation = mockDataSource?.locations?[0] else { return }
+        mockDataSource?.didUpdate(initialLocation)
+    }
+    
+    /// Called in response to the Navigate button being tapped.
+    @IBAction func startNavigation() {
+        navigateButtonItem.isEnabled = false
+        resetButtonItem.isEnabled = true
+        mapView.locationDisplay.autoPanMode = .navigation
+        // If the user navigates the map view away from the location display, activate the recenter button.
+        mapView.locationDisplay.autoPanModeChangedHandler = { [weak self] _ in self?.recenterButtonItem.isEnabled = true }
+        // Start the location data source and location display.
+        mapView.locationDisplay.start(completion: nil)
+    }
+    
+    /// Called in response to the Reset button being tapped.
+    @IBAction func reset() {
+        // Stop the speech, if there is any.
+        speechSynthesizer?.stopSpeaking(at: .immediate)
+        speechSynthesizer = nil
+        resetToStartingLocation()
+        // Stop the datasource generation, if there is any.
+        mapView.locationDisplay.stop()
+        mapView.locationDisplay.autoPanMode = .off
+        graphicsOverlay.graphics.removeAllObjects()
+        directionsList = nil
+        // Reset the navigation.
+        setDataSource(routeResult)
+        setRouteTracker(routeResult)
+        setRouteGraphics(routeResult)
+        setSpeechSynthesizer()
+        recenterButtonItem.isEnabled = false
+        resetButtonItem.isEnabled = false
+        navigateButtonItem.isEnabled = true
+    }
+    
+    /// Called in response to the Recenter button being tapped.
+    @IBAction func recenter() {
+        mapView.locationDisplay.autoPanMode = .navigation
+        recenterButtonItem.isEnabled = false
+    }
+    
+    // MARK: UIViewController
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        // Add the source code button item to the right of navigation bar.
+        (self.navigationItem.rightBarButtonItem as? SourceCodeBarButtonItem)?.filenames = ["NavigateRouteViewController"]
+    }
+    
+    override func viewWillDisappear(_ animated: Bool) {
+        // Nullify the properties to avoid lingering tasks running in the background.
+        super.viewWillDisappear(animated)
+        defaultViewPoint = nil
+        directionsList = nil
+        routeTracker = nil
+        mockDataSource = nil
+        speechSynthesizer = nil
```

Comment:
I don't think any of these should be unset when the view disappears. If the view disappears because the user opens the readme, all of these things would need to be recalculated/recreated again. If the view disappears because the sample is closed, as you pointed out, the view controller will get deallocated when another sample is opened.